### PR TITLE
[WIP] Prefix message signing according to eth_sign. (#575)

### PR DESCRIPTION
**Pull request by ricmoo**
As we discussed before, I have been unable to run the ./gradlew integration tests. But opening a PR so it is easier to get this discussion going. :)

I've updated the test cases by using the key pair to sign a message using ethers. I will be adding more test cases to ethers soon, generated by parity and geth.

I am also concerned that any dapp that has previously signed a message with web3j will no longer get the correct key back when calling `signedMessageToKey`, so this is at best, something that should probably be a major version change. The other option is to add a `signedPrefixedMessageToKey` and a `signPrefixedMessage` and deprecate the non-prefixed version.

### CODE REVIEW

**Comment by ricmoo** [comment]
Ah. After Travis CI, I now see how the needToHash works... I think I have an idea, one sec. :)

**Comment by ricmoo** [comment]
There are now separate messages for `signMessage` and `signPrefixedMessages` (and `signedMessageToKey` and `signedPrefixedMessageToKey`). This should preserve backward compatibility, since I see the `signMessage` is used for signing transaction RLP.

The documentation should point out that `signPrefixedMessage` should be used to emulate signing messages (e.g. `eth_sign`).

**Comment by conor10** [comment]
Thanks @ricmoo. So should `signPrefixedMessages` only be used in `eth_sign`, not in transactions signed by web3j?

**Comment by ricmoo** [comment]
Currently the signPrefixedMessage does the same thing as [eth_sign](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_sign) for messages, so it can be used in  lieu of calling the eth_sign method of a Geth node, if you have the private key.

The current signMessage, is what signing transactions requires.

Ideally we would do the following:

1. Rename signMessage to signDigest (and pass in the Hash.sha3 instead of compute it inside)
2. Rename signPrefixedMessage to signMessage

But that would cause some non-backwards compatible changes. Maybe something that can be rolled into your next major release?

**Comment by snazha-blkio** [comment]
I've changed the target branch to 4.0, as recommended by @iikirilov

**Comment by ricmoo** [comment]
I agree with that plan. :)

File: crypto/src/main/java/org/web3j/crypto/Sign.java [hunk 1]
```diff
@@ -36,6 +36,26 @@
             CURVE_PARAMS.getCurve(), CURVE_PARAMS.getG(), CURVE_PARAMS.getN(), CURVE_PARAMS.getH());
     static final BigInteger HALF_CURVE_ORDER = CURVE_PARAMS.getN().shiftRight(1);
 
+    static final String MessagePrefix = "\u0019Ethereum Signed Message:\n";
```
Comment (conor10) [inline crypto/src/main/java/org/web3j/crypto/Sign.java]: This should be MESSAGE_PREFIX

File: crypto/src/main/java/org/web3j/crypto/Sign.java [hunk 2]
```diff
@@ -36,6 +36,26 @@
...
+
+    static byte[] getMessagePrefix(int messageLength) {
+        return MessagePrefix.concat(String.valueOf(messageLength)).getBytes();
+    }
+
+    static byte[] getMessageHash(byte[] message) {
```
Comment (conor10) [inline crypto/src/main/java/org/web3j/crypto/Sign.java]: Should we call this `etEthereumMessageHash` instead? As it is an Ethereum specific method for hashing messages.

Comment (snazha-blkio) [reply crypto/src/main/java/org/web3j/crypto/Sign.java]: Address in: https://github.com/web3j/web3j/pull/761/

File: crypto/src/main/java/org/web3j/crypto/Sign.java (hunk 1 above)
Comment (snazha-blkio) [reply crypto/src/main/java/org/web3j/crypto/Sign.java]: Addressed in https://github.com/web3j/web3j/pull/761/

**Comment by snazha-blkio** [comment]
Hi @ricmoo 
We have created a separate branch from this fork to address some PR feedback, we have now merged the changes into the release/4.0 branch
c.f. https://github.com/web3j/web3j/pull/761

Thanks for the contribution, great work!

**Pull request merged**